Addition of Global Lightning Probability Module Based on USAF Algorithm

timbonin opened the pull request:
Adding the lightning algorithm that MIT LL has added developed based on current USAF documentation of their operational algorithm used in house. There will be a companion PR going up shortly for acceptance tests accompany this.

I spent a fair amount of time making sure the code style and all the other checks passed, so that should make this a bit easier to review. Open to any suggestions on if things should be reformatted.  We added the logic of the code to the existing lightning.py file as we thought it made sense, but have a separate CLI call for invoking the algorithm 'lightning_multivariate_probability'.

Let me know if you have any comments/concerns.

Acceptance test data PR: https://github.com/metoppv/improver_test_data/pull/44

Testing:
 - [X] Ran tests and they passed OK
 - [X] Added new tests for the new feature(s)

CLA
 - [X] If a new developer, signed up to CLA

ivorblockley commented:
This would be a great addition to IMPROVER, and very interesting to see what diagnostic schemes are in use elsewhere.

At the Australian Bureau of Meteorology we have an alternative algorithm that makes use of the instantaneous cloud physics thunder parameter (Bright et al. 2005) along with accumulated precipitation, although we are in the process of developing improved formulations using additional convective parameters and we may also implement this eventually in the IMPROVER framework. As such, I would request we may be more careful in naming the Python modules or functions/classes here to make them a little less generic, e.g. `class LightningMultivariateProbability_USAF2024` or similar

gavinevans commented:
Thanks @ivorblockley. We'll try to have a conversation within the Met Office regarding the naming convention issue that you raise here, just to try to come up with a recommendation, as this could potentially impact other algorithms already within the IMPROVER codebase.

timbonin commented:
I am open to changing the name of the module to whatever we decide is best going forward. To be honest, our working name for this internally was to call it LightningProbability_USAF (without the year), but  after reviewing the other names we decided to change the name as no other ones were named specific to any one agency.  Whatever we decide as a group is best, we are happy to go with.

gavinevans commented on improver/lightning.py:
```diff
@@ -204,3 +204,220 @@ def latitude_to_threshold(
+        # Make percentage:
+        data = lprob * 100.0
+
+        cube = create_new_diagnostic_cube(
+            name="""20_km_lightning_probability_over_the_valid_time_of_the_accumulated_
+                precipitation""",
+            units="1",
+            template_cube=templ,
+            data=data.astype(FLOAT_DTYPE),
+            mandatory_attributes=generate_mandatory_attributes(
+                cubes, model_id_attr=model_id_attr
+            ),
+        )
```
I think that I've got a few metadata queries.

1. I note that this output, following the precipitation accumulation input, has a time coordinate where the validity time is defined as a midpoint between the bounds. For period diagnostics, we always define the validity time to match the end bound, so I thought that I'd query this. Is the use of validity time as the midpoint for period diagnostics your preferred approach? I note that the forecast period also uses the midpoint of the bounds as the "point".
2. Currently this cube is a probability cube but doesn't have a threshold coordinate. Would it be possible to add a threshold coordinate, similar to the `number_of_lightning_flashes per_unit_area` coordinate in the `LightningFromCapePrecip` plugin? Perhaps something like `lightning_in_area` would work as a threshold coordinate name? Currently, this cube is a bit implicit about the fact that it is predicting the probability of _any_ lightning. 
3. The cube name is a bit overloaded at the moment. Our lightning probability diagnostic is named: "probability_of_number_of_lightning_flashes_per_unit_area_in_vicinity_above_threshold", so perhaps something like `probability_of_lightning_in_vicinity_above_threshold` would be better. As our lightning probability diagnostic is also within a particular area, we've added a `radius_of_vicinity` coordinate, which in your case would be `radius_of_vicinity: 20000 m`. Would it make sense to arrange the metadata like that from your perspective?

timbonin replied:
With regard to the above comments:

1. I was unaware of the convention that the Met Office followed regarding validity time falling at the end of a period for diagnostics.  We don't have a preferred approach and are fine with either method.  This algorithm is written in a way so that the convention used for the accumulated precipitation input will be propagated through to the outputs, so it really should support either convention of the input data.  Regardless, I've updated the unit tests to follow the Met Office convention of having the validity time be at the end of the period. 
2. Yes, we can add a threshold coordinate. Given that the `LightningFromCapePrecip` plugin also is forecasting the probability of _any_ lightning, I'm going to use the exact same threshold coordinate name as what that plugin uses for consistency.
3. Got it.  I simplified the name as you suggested to `probability_of_lightning_in_vicinity_above_threshold` and added an auxiliary coordinate to note the vicinity over which this product is valid for 20 km.

gavinevans commented on improver/lightning.py:
```diff
@@ -204,3 +204,220 @@ def latitude_to_threshold(
+
+    ---------------------------+------------+----------------------------------------
+
+    Regression equation when CAPE and APCP are greater than zero:
```
Just to ask, is this algorithm available anywhere in a referenceable location? If so, it would be great to add the reference.

timbonin replied:
This actually came up as a question from the BoM as well when I had a call with them earlier this week.  Unfortunately, there is no reference available as this came from internal documentation that we received from the USAF about the algorithms they run in house in their current post-processing system. This algorithm had been developed from some previous statistical analysis the USAF had conducted.

gavinevans commented on improver/lightning.py:
```diff
@@ -204,3 +204,220 @@ def latitude_to_threshold(
     )
+
+
+class LightningMultivariateProbability(PostProcessingPlugin):
```
This isn't a definitive recommendation but related to [this comment](https://github.com/metoppv/improver/pull/1983#issuecomment-1992736096), I think that it would be reasonable to name this `LightningMultivariateProbability_USAF2024` where USAF2024 is effectively the reference in lieu of a citable reference. Therefore, I would think that an algorithm taken from e.g. Bright et al., 2005 might be named `LightningMultivariateProbability_Bright2005`. I think that this naming convention is OK for now, although we're planning to have some more internal discussions about this.

timbonin replied:
I'm ok with this renaming and have changed the name throughout the codebase.

gavinevans commented on improver_tests/acceptance/test_lightning_multivariate_probability.py:
As you've added a CLI and new acceptance test data, this PR should include an update to [this file](https://github.com/metoppv/improver/blob/master/improver_tests/acceptance/SHA256SUMS). This file should be updated following the instructions in [ReadTheDocs](https://improver.readthedocs.io/en/latest/How-to-implement-a-command-line-utility.html#updating-acceptance-test-data-checksums). Please let us know if this documentation isn't clear.

timbonin replied:
Got it, thanks for the guidance. I've updated the checksums file and ensured our KGO passes through acceptance tests on our end with it.

gavinevans reviewed (changes requested):
Thanks for the PR. This will be a useful addition to the codebase @timbonin 👍 

I've added some comments below.

timbonin commented on improver/lightning.py:
```diff
@@ -204,3 +204,220 @@ def latitude_to_threshold(
+    Outputs:
+    20 km lightning probability over the valid time of the accumulated precipitation
+
+    ---------------------------+------------+----------------------------------------
```
I honestly don't know how the pluses got in that line, and how we didn't catch that previously.  Yes, we'll change it as you suggest.

timbonin commented on improver/lightning.py:
```diff
@@ -204,3 +204,220 @@ def latitude_to_threshold(
+    Lifted Index (liftidx)
+    Precipitable Water (pwat in mm)
+    CIN (cin)
+    3-hour Accumulated Precipitation (apcp in inches)
```
I agree that we should clarify the units we're expecting for inputs, as well as acronyms and have now specified both as suggested.

timbonin commented on improver/lightning.py:
```diff
@@ -204,3 +204,220 @@ def latitude_to_threshold(
+        if apcp:
+            apcp = apcp.merge_cube()
+        else:
+            raise ValueError(f"No cube named precipitation_amount found in {cubes}")
```
Good idea to simplify the code and redundancy. I've made this change.

timbonin commented on improver/lightning.py:
```diff
@@ -204,3 +204,220 @@ def latitude_to_threshold(
+
+        (cape_time,) = list(cape.coord("time").cells())
+        (apcp_time,) = list(apcp.coord("time").cells())
+        if cape_time.point != apcp_time.bound[0]:
+            raise ValueError(
+                f"CAPE cube time ({cape_time.point}) should be valid at the "
+                f"precipitation_accumulation cube lower bound ({apcp_time.bound[0]})."
+            )
```
Yes, this requirement should be done for the other cubes as well (CAPE, lifted index, CIN, and precipitable water).  Modified this check as you suggested to loop over the cubes, and have added the following docstring in the CLI to make this clear:

`
    The cubes for CAPE, lifted index, precipitable water, and CIN must be valid for the beginning
    of the 3-hr accumulated precipitation window.
`

timbonin commented on improver/lightning.py:
```diff
@@ -204,3 +204,220 @@ def latitude_to_threshold(
+                f"Precipitation_accumulation cube time window must be three hours, "
+                f"not {np.diff(apcp_time.bound)}."
+            )
+        # Following time/space checks should be made for all diagnostics
```
Yes, good catch. It should be made for all diagnostics. I actually meant to add this a few weeks ago before checking in this code, but forgot. Made change as suggested.

timbonin commented on improver/lightning.py:
```diff
@@ -204,3 +204,220 @@ def latitude_to_threshold(
+            raise ValueError(
+                "Supplied cubes must have the same forecast reference times"
+            )
+        if not spatial_coords_match([cape, apcp]):
+            raise ValueError("Supplied cubes do not have the same spatial coordinates")
```
Yes, similarly to the last one this should be checked on all cubes. Changed as suggested and updated unit tests as appropriate.

timbonin commented on improver/lightning.py:
```diff
@@ -204,3 +204,220 @@ def latitude_to_threshold(
+
+    def process(self, cubes: CubeList, model_id_attr: str = None) -> Cube:
+        """
+        From the supplied CAPE, LIFTIDX, PWAT, CIN, APCP cubes, calculate a probability
+        of lightning cube.
+
+        Args:
+            cubes:
+                Cubes of CAPE, LIFTIDX, PWAT, CIN, APCP.
```
Yes, spelled these out in the docstring as suggested. Did not spell out CAPE and CIN, which are common acronyms and CAPE is not spelled out in existing lightning algorithm from CAPE and Precipitation Rate.

timbonin commented on improver/lightning.py:
```diff
@@ -204,3 +204,220 @@ def latitude_to_threshold(
+        # access the actual instability in very moist environments:
+        lprob_noprecip = cape / (cin + 100.0)
+        lprob_noprecip = 0.025 * np.log(lprob_noprecip + 0.31) + 0.03
+        apcp = apcp - (pwat / 1000)
```
This does effectively convert `pwat` to `m` for this calculation. Thus, we are indeed subtracting precipitation accumulation in inches from precipitable water in meters.  However, this whole regression equation is statistical in general and does not really have any physical meaning, thus this is incidental.

timbonin commented on improver_tests/lightning/test_LightningMultivariateProbability.py:
```diff
@@ -0,0 +1,274 @@
+    return cube
+
+
+def test_basic(cape_cube, liftidx_cube, pwat_cube, cin_cube, apcp_cube, expected_cube):
```
Good catch. In my reworking of the unit tests, I've reworked and manually put in a range of values for each of the parameters (CAPE, CIN, precipitation accumulation, lifted index, and precipitable water) in the numpy arrays so that all the different conditions in the regression equations are now covered to ensure the expected output is produced for the wide range of conditions.

Specifically, instead of setting each of the arrays to fill values for the entire cube I've set the values to vary within each cube fixture, such as for CAPE:
`    data = np.array([[4000, 0], [0, 4000]], dtype=np.float32)`

timbonin commented:
Thank you @gavinevans  for the thorough review and helping me walk through the process of my first IMPROVER PR.  I think I've learned from this experience and hopefully subsequent ones will be smoother, as I'll be aware of the common issues raised during this one (e..g, comprehensive unit tests, checksums, documentation, etc).  

I've 'resolved' all the conversations mostly so that I could track which ones I've addressed. Feel free to review them and make sure I've adequately addressed all the comments/concerns and let me know if you have any other suggestions.

gavinevans commented on improver/lightning.py:
```diff
@@ -204,3 +205,217 @@ def latitude_to_threshold(
+        lprob[lprob > 0.95] = 0.95
+
+        # Make percentage:
+        data = lprob * 100.0
```
I think that we usually keep probabilities as values between 0 and 1, as in the `LightningFromCapePrecip` plugin. Is there is reason to make this a percentage here?

timbonin replied:
We did this simply to make it a percentage as that is how the USAF wants it shown. However, I can (and have) easily removed this and keep to keep it fractional to be aligned with the IMPROVER convention, and we can simply magnify it later for display.

gavinevans commented on improver_tests/lightning/test_LightningMultivariateProbability_USAF2024.py:
```diff
@@ -0,0 +1,270 @@
+    result = LightningMultivariateProbability_USAF2024()(
+        CubeList([cape_cube, liftidx_cube, pwat_cube, cin_cube, apcp_cube]), None
+    )
+    assert np.allclose(result.data, expected_cube.data)
```
It would be a good idea to add an assertion here to check some of the metadata. The relevant metadata things to check might be the threshold coordinate, the vicinity coordinate and checking that the time point and bound of the expected cube matches the precipitation accumulation cube. One option would be to use the lines below that are used in the `LightningFromCapePrecip` unit tests:

```
    assert result.xml().splitlines(keepends=True) == expected_cube.xml().splitlines(
        keepends=True
    )
```

timbonin replied:
Ok, I've added these lines to the test to check the metadata is correct; I've also added some other metadata to the expected cube as necessary.

gavinevans commented on improver_tests/acceptance/test_lightning_multivariate_probability_usaf2024.py:
```diff
@@ -42,7 +42,7 @@
 def test_basic(tmp_path, with_model_attr):
     """Test basic invocation"""
-    kgo_dir = acc.kgo_root() / "lightning-multivariate-probability"
+    kgo_dir = acc.kgo_root() / " lightning-multivariate-probability-usaf2024"
```
```suggestion
    kgo_dir = acc.kgo_root() / "lightning-multivariate-probability-usaf2024"
```

gavinevans reviewed (changes requested):
Thanks for the updates, @timbonin 👍 

I've added a few more minor comments.

timbonin commented on improver/lightning.py:
```diff
@@ -204,3 +205,217 @@ def latitude_to_threshold(
+
+        cape, liftidx, pwat, cin, apcp = output_cubes
+
+        (cape_time,) = list(cape.coord("time").cells())
```
You're right, this is a residual from a previous version of the code prior to the additional checks we have inserted wherein the valid times are checked for all the cubes. I've removed this line.

timbonin commented:
Ok, I think we've addressed all the comments and concerns! I've updated the unit tests for metadata, created acceptance tests that are not idealized that, ensured the validity time was at the end of the period bounds, and made the additional minor changes.  

Further, we did make a few minor changes to the algorithm (e.g., we changed the sign on CIN), as conventionally CIN is given as <=0 whereas the regression equations used assumed it was >=0.  We realized we were doing a pre-processing step on our CIN values to reverse the convention, but we've moved this reversal directly into the algorithm so that any users can pass in CIN cubes as they are standardly defined.

gavinevans commented:
Thanks for all the updates @timbonin 👍 . I've added one query to the associated test data PR: https://github.com/metoppv/improver_test_data/pull/44, but otherwise this PR looks fine to me now.

timbonin commented:
> Thanks for all the updates @timbonin 👍 . I've added one query to the associated test data PR: [metoppv/improver_test_data#44](https://github.com/metoppv/improver_test_data/pull/44), but otherwise this PR looks fine to me now.

I've just addressed the other query on the acceptance test data, so I think we should be good to go.  I'll double check with you one more time.

gavinevans reviewed (dismissed):
Thanks for all the updates @timbonin 👍 . This looks fine to me now. 

I'll get someone else to give this a second review.

bayliffe commented on improver/lightning.py:
```diff
@@ -204,3 +205,215 @@ def latitude_to_threshold(
+        """
+
+        output_cubes = iris.cube.CubeList()
+        input_names = [
+            "atmosphere_specific_convective_available_potential_energy",
+            "temperature_difference_between_ambient_air_and_air_lifted_adiabatically",
+            "precipitable_water",
+            "atmosphere_specific_convective_inhibition",
+            "precipitation_amount",
+        ]
+
+        for input_name in input_names:
+            output_cubes.append(self._extract_input(cubes, input_name))
```
```suggestion
        input_names = {
            "atmosphere_specific_convective_available_potential_energy": ["J kg-1"],
            "temperature_difference_between_ambient_air_and_air_lifted_adiabatically": ["K"],
            "precipitable_water": ["kg m-2", "mm"],
            "atmosphere_specific_convective_inhibition": ["J kg-1"],
            "precipitation_amount": ["kg m-2", "mm"],
        }

        for input_name, units in input_names.items():
            output_cubes.append(self._extract_input(cubes, input_name))
            if not output_cubes[-1].units in units:
                raise ValueError("Input diagnostic units are incorrect")
```

Can I suggest something along these lines to check that the inputs units are what we expect, and if not raise an exception? It seems particularly important here as the units are stripped later on so we need to ensure they're correct whilst we can.

bayliffe replied:
This would probably want a better exception message than my rather vague suggestion.

bayliffe commented on improver/lightning.py:
```diff
@@ -204,3 +205,215 @@ def latitude_to_threshold(
+    values of lifted index are positive here:
+    LIFTIDX=LIFTIDX+4.0
+    IF LIFTIDX is less than 0 THEN LIFTIDX=0.0
+    IF CAPE is less than 0 THEN lprob =0.2*(LIFTIDX*APCP)^0.5
```
I'm getting pretty picky, but below you define it as `<=`, so either here or there we should make these the same.

timbonin replied:
Got it, defined it as <= here for CAPE to match the operations we actually perform in the code below.

bayliffe commented on improver/lightning.py:
```diff
@@ -204,3 +205,215 @@ def latitude_to_threshold(
+
+        # Finally, the probability of lightning is reduced when there is not much PWAT, since
+        # graupel cannot form required for the charging process. The probability is reduced as:
+        lprob[np.where(pwat < 20)] = lprob[np.where(pwat < 20)] * (
+            pwat[np.where(pwat < 20)] / 20.0
+        )
```
```suggestion
        low_pwat = np.where(pwat < 20)
        lprob[low_pwat] = lprob[low_pwat] * (pwat[low_pwat] / 20.0)
```
This change will give you a fractional speed up by reducing the number of tests.

bayliffe commented on improver/lightning.py:
```diff
@@ -204,3 +205,215 @@ def latitude_to_threshold(
+        Returns:
+            Cube of lightning data
+
+        Raises:
```
A pure stylistic thing is that we would usually put the raises bit of the doc-string into the method that does the raising, so in this case, the `_get_inputs` doc-string. However, as it's a private method you might be doing things a bit differently, therefore feel free to entirely ignore me if you wish.

timbonin replied:
We modeled our `LightningMultivariateProbability_USAF2024` plug-in off of the existing `LightningFromCapePrecip` plug-in, which has the raise doc-strings in a similar location.  To be consistent with the existing module, we're going to leave it as is.

bayliffe commented on improver_tests/lightning/test_LightningMultivariateProbability_USAF2024.py:
```diff
@@ -0,0 +1,297 @@
+    with pytest.raises(ValueError, match=error_msg):
+        LightningMultivariateProbability_USAF2024()(
+            CubeList([cape_cube, apcp_cube, cin_cube, liftidx_cube, pwat_cube])
+        )
```
If you adopt my suggestion to check the input units, you can also test breaking those here as well.

timbonin replied:
Added a check into the unit tests to ensure that the units are checked for consistency.

bayliffe reviewed (changes requested):
@timbonin I've added some suggestions, none of which are vital, though I would emphasise the units checking as useful. Feel free to argue with any of this.

ivorblockley reviewed (commented):
I made a number of review comments a few weeks ago but forgot to "submit" the pending review. My apologies! I've deleted some old comments that pertain to things that have since been addressed, nonetheless it is likely some aspects of the comments are now stale.

timbonin commented on improver/lightning.py:
```diff
@@ -204,3 +205,215 @@ def latitude_to_threshold(
+    lprob_noprecip=0.025*alog(lprob_noprecip+0.31)+0.03
+    APCP=APCP-(PWAT/1000)
+
+    IF APCP is less than 0.01 THEN lprob=lprob_noprecip
```
Added units here (at line 241) and elsewhere where I think it's appropriate to clarify.

timbonin commented on improver/lightning.py:
```diff
@@ -204,3 +205,215 @@ def latitude_to_threshold(
+
+    Finally, the probability of lightning is reduced when there is not much PW, because graupel
+    cannot form and start the whole charging process. Therefore we reduce the probability:
+    IF PWAT is less than 20 THEN lprob=lprob*(PWAT/20.0)
```
Added units. I wish they were as interesting as fleas in a pyramidal structure on a tiny motorbike. 😃

timbonin commented on improver_tests/acceptance/test_lightning_multivariate_probability_usaf2024.py:
```diff
@@ -0,0 +1,65 @@
+# -*- coding: utf-8 -*-
```
That is weird and I'm not honestly quite sure how those permissions were set like that. It must be residual from when we were working through some issues in house.  Anyway, I've corrected this and also realized the permissions on the unit tests were set similarly and fixed those as well.

timbonin commented on improver/lightning.py:
```diff
@@ -204,3 +204,220 @@ def latitude_to_threshold(
+    the model often produces showery precipitation that doesn’t access the actual instability in
+    very moist environments:
+    lprob_noprecip=CAPE/(CIN+100.0)
+    lprob_noprecip=0.025*alog(lprob_noprecip+0.31)+0.03
```
Changed as suggested in the two locations you point out. Good catch!

```
    Regression equation when CAPE and APCP are greater than zero:
    lprob= cape*APCP
    lprob=0.13*ln(lprob+0.7)+0.05

    If APCP is very low, a separate regression equation is used to predict lightning probability.
    The definition of “very low” is raised slightly when PWAT values are high. This is because
    the model often produces showery precipitation that doesn’t access the actual instability in
    very moist environments:
    lprob_noprecip=CAPE/(CIN+100.0)
    lprob_noprecip=0.025*ln(lprob_noprecip+0.31)+0.03
    APCP=APCP-(PWAT/1000)
```

timbonin commented on improver/lightning.py:
```diff
@@ -204,3 +204,220 @@ def latitude_to_threshold(
+        liftidx = liftidx.data
+        pwat = pwat.data
+        cin = cin.data
+        apcp = apcp.data / 1000 * 39.3701  # convert (kg m-3) to (inches)
```
Yes, Ben raised the same comment above with regard to checking for units.  I don't necessarily want to convert the units within the code itself, especially if an input cube has completely incompatible units for some reason it may raise other issues that need to be resolved. However, I agree we should definitely check the units to make sure they're what we expect for the regression statistics to be applicable. 

We have added the following check in the `_get_inputs()` which does as you (and Ben) suggest, and I've also added a unit test to make sure this works: 

```
        output_cubes = iris.cube.CubeList()
        input_names = {
            "atmosphere_specific_convective_available_potential_energy": ["J kg-1"],
            "temperature_difference_between_ambient_air_and_air_lifted_adiabatically": ["K"],
            "precipitable_water": ["kg m-2", "mm"],
            "atmosphere_specific_convective_inhibition": ["J kg-1"],
            "precipitation_amount": ["kg m-2", "mm"],
        }

        for input_name, units in input_names.items():
            output_cubes.append(self._extract_input(cubes, input_name))
            if not output_cubes[-1].units in units:
                expected_unit_string = ' or '.join(map(str, units))
                received_unit_string = str(output_cubes[-1].units)
                raise ValueError(
                    f"The {output_cubes[-1].name()} units are incorrect, expected "
                    f"units as {expected_unit_string} but received {received_unit_string})."
```

ivorblockley replied:
perfect, great to have these checks, thank you

timbonin commented on improver/lightning.py:
```diff
@@ -204,3 +204,220 @@ def latitude_to_threshold(
+    CIN (cin)
+    3-hour Accumulated Precipitation (apcp in inches)
+    Outputs:
+    20 km lightning probability over the valid time of the accumulated precipitation
```
Ideally, we would use a parcel that's lifted from the most unstable layer (which is what these statistics were developed from). When possible, that is what we use. However, there are some models that we've used that do not provide MUCAPE but instead we get surface based.  In those circumstances, we use the surface CAPE and the results tend to be very similar, except we've noticed that there are times and conditions when there's elevated convection that's missed in the lightning forecast.  We've clarified that here:
```
    Convective Available Potential Energy (CAPE in J/kg. Most Unstable CAPE was used to determine
    the regression statistics, but surface-based CAPE is supported and yields similar results but 
    will not forecast elevated convection.)
```

Clarified lifted index here to be between parcel and ambient air temperature at 500 mb:
`    Lifted Index (liftind in K, difference between parcel and ambient temperatures at 500 mb)`

The last comment about `acpc` being the incorrect units is stale, as we have corrected this in the previous round of comments.
`    3-hour Accumulated Precipitation (apcp in kg m-2 or millimetres)`

ivorblockley replied:
thank you, those doc updates make it much clearer and provide useful additional context

timbonin commented on improver/lightning.py:
```diff
@@ -204,3 +204,220 @@ def latitude_to_threshold(
+        liftidx[liftidx < 0] = 0
+
+        lprob[np.where(cape <= 0)] = (
+            0.2 * (liftidx[np.where(cape <= 0)] * apcp[np.where(cape <= 0)]) ** 0.5
```
With regard to the two points:

1. You're correct in that `acpc` was modified earlier, and we realized that it was an error in our code as it should not have carried through to this equation.  We have corrected this as we did find it led to some errors that you mentioned in certain circumstances. Here's the snippet of code showing how we changed it so that we created a temporary copy of acpc where it is modified:
```
        apcp_temp = apcp - (pwat / 1000)

        lprob[np.where(apcp_temp < 0.01)] = lprob_noprecip[np.where(apcp_temp < 0.01)]

        # If there is no CAPE but the atmosphere is “close” to unstable, lightning does sometimes
        # occur, especially when heavy precipitation may have stabilized the atmosphere in the
        # model. Unstable values of lifted index are positive here:
        liftidx = liftidx + 4.0

        liftidx[liftidx < 0] = 0

        lprob[np.where(cape <= 0)] = (
            0.2 * (liftidx[np.where(cape <= 0)] * apcp[np.where(cape <= 0)]) ** 0.5
        )
```

2. I agree that this is not an intuitive result, but we've double checked the USAF regression algorithms and the documentation we have and this is in line with the statistics from the analysis that they have conducted.

ivorblockley replied:
1. Indeed, creating the `apcp_temp` temporary array solves this neatly.

2. Thanks for double-checking. It is curious but I agree about faithfully implementing the approach the USAF developed here

ivorblockley commented:
Thanks @timbonin for addressing all my feedback.

bayliffe reviewed (approved):
Thanks for the changes Tim, I'm happy with this now.
